### PR TITLE
Faster slice PartialOrd

### PR DESCRIPTION
**Pull request by ranma42**
This branch improves the performance of Ord and PartialOrd methods for slices compared to the iter-based implementation.
Based on the approach used in #26884.

### CODE REVIEW

**Comment by ranma42** [comment]
Benchmarking data is based on https://gist.github.com/ranma42/51d46e4a4664e388a8f7

File: src/libcore/slice.rs [hunk 1]
```diff
@@ -1558,8 +1558,18 @@ impl<T: Eq> Eq for [T] {}
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: Ord> Ord for [T] {
+    #[inline]
     fn cmp(&self, other: &[T]) -> Ordering {
-        self.iter().cmp(other.iter())
+        let l = cmp::min(self.len(), other.len());
+
+        for i in 0..l {
+            match self[i].cmp(&other[i]) {
```
Comment (bluss) [inline src/libcore/slice.rs]: Are bounds checks eliminated here? [This](https://users.rust-lang.org/t/how-to-zip-two-slices-efficiently/2048) suggests you need to slice and other to length `l` for it to be.

Comment (bluss) [inline src/libcore/slice.rs]: (I guess this wasn't needed in the partialeq case since it requires equal lengths, and llvm understands that)

Comment (ranma42) [inline src/libcore/slice.rs]: You are right, in order to remove all bound checks I need to re-slice `self` and `other`, which further improves performance :)

Comment (bluss) [inline src/libcore/slice.rs]: Yay!

**Comment by ranma42** [comment]
@bluss In a previous version my code was using Some(self.cmp(other)), but I was not sure about which form was preferred. I will rewrite as you suggested.

File: src/libcore/slice.rs (hunk 1 above)
Comment (bluss) [inline src/libcore/slice.rs]: When you add this, maybe add a comment to indicate why you reslice.

**Comment by bluss** [comment]
I almost think the main win is not using the iterator comparators' 1) `.zip()` and 2) match on tuples, both which @dotdash pointed out have far from optimal codegen.

**Comment by ranma42** [comment]
@bluss Another significant win (which kind of surprised me) was not specialising the boolean comparisons "eagerly". It is faster to only check for equality in the loop and then perform the dispatch "outside" than to have an explicit implementation like
```Rust
    #[inline]
    fn le(&self, other: &[T]) -> bool {
        let l = cmp::min(self.len(), other.len());
        let lhs = &self[..l];
        let rhs = &other[..l];

        for i in 0..l {
            match lhs[i].partial_cmp(&rhs[i]) {
                Some(Ordering::Equal) => (),
                Some(Ordering::Less) => return true, // this
                _ => return false,  // makes the loop slower
            }
        }

        self.len().le(&other.len())
    }
```

**Comment by bluss** [comment]
Oh that's cool, thanks for the explanation!

**Comment by ranma42** [comment]
Updated with improvements suggested by @bluss (and new benchmarking results).
As expected, the main part of the loop (which scans the slice looking for elements which are not equal) runs about as fast as the `eq` and `ne` implementation.
The commit messages provide some insight about how each commit affects performance and/or the results of codegen (the bounds check optimisation is mentioned in the fourth commit; the equals-only check is mentioned in the second one). Would it be better to mention it directly in the code (as comments)?

**Comment by bluss** [comment]
Yes, I think it's best to leave a comment for the reslicing. It looks redundant for the reader.

**Comment by ranma42** [comment]
It might also be a good idea to avoid declaring `cmp` and `partial_cmp` as inline (and it would be consistent with the `eq` and `ne` implementation). Any opitions on this?

**Comment by bluss** [comment]
They don't need inline, since the trait impl is already generic. I think it's fine to remove.. we are trying to be conservative with the inlines.

**Comment by bluss** [comment]
Thank you, it looks good! Great benchmark wins!

@bors r+

**Pull request merged**